SolarTracker/Pages/Index.razor

```diff
@@ -0,0 +1,362 @@
+        int day = int.Parse(dateSegments[2]);
+        _selectedDateTime = new DateTime(year, month, day, _selectedDateTime.Hour, _selectedDateTime.Minute, 0);
+        _terminusAdded = false;
+        await OnMapRendered();
+    }
+
+    private async Task UpdateTime(ChangeEventArgs arg)
+    {
+        string[]? timeSegments = arg.Value?.ToString()?.Split(':');
+        if (timeSegments is null || timeSegments.Length < 2) return;
+        int hour = int.Parse(timeSegments[0]);
+        int minutes = int.Parse(timeSegments[1]);
+        _selectedDateTime = new DateTime(_selectedDateTime.Year, _selectedDateTime.Month, _selectedDateTime.Day, hour, minutes, 0);
+        _terminusAdded = false;
+        await OnMapRendered();
+    }
+    
+    private async Task UpdateTimeZone(ChangeEventArgs arg)
+    {
+        _selectedDateTime = _selectedDateTime.Add(-_timeZoneOffset);
+        string? timeZoneId = arg.Value?.ToString();
+        if (timeZoneId is null) return;
+        _timeZoneInfo = _timeZoneInfos.FirstOrDefault(tzi => tzi.Id == timeZoneId);
+        if (_timeZoneInfo is null) return;
+        _timeZoneOffset = _timeZoneInfo.GetUtcOffset(_selectedDateTime);
+        _selectedDateTime = _selectedDateTime.Add(_timeZoneOffset);
+    }
+
+    private async Task OnExtentChanged()
+    {
+        await GetCenterAndTimeZone();
+    }
+
+    private async Task GoToSunrise()
+    {
+        _movingToSunrise = true;
+        _selectedDateTime = _selectedDateTime.AddMinutes(_skipInterval);
+        await OnMapRendered();
+    }
+
+    private async Task GoToSunset()
+    {
+        _movingToSunset = true;
+        _selectedDateTime = _selectedDateTime.AddMinutes(_skipInterval);
+        await OnMapRendered();
+    }
+
+    private async Task MoveTimeUntilStateChanged(int minutesInterval, bool forward, bool isIntersected)
+    {
+        _selectedDateTime = _selectedDateTime.AddMinutes(forward ? minutesInterval : -minutesInterval);
+        await CreateTerminator();
+        bool nextDirection = forward;
+        bool nextIntersected = isIntersected;
+        if (isIntersected != await GeometryEngine.Intersects(_center!, _polygon!))
+        {
+            nextDirection = !forward;
+            nextIntersected = !isIntersected;
+        }
+
+        if (minutesInterval > 1)
+        {
+            await MoveTimeUntilStateChanged((int)(minutesInterval / 2.0), nextDirection, nextIntersected);
+        }
+    }
+
+    private async Task FetchTimeZoneData()
+    {
+        try
+        {
+            var client = new HttpClient();
+            HttpResponseMessage result;
+            
+            List<Task> fetchTasks = new();
+            for (int i = 1; i < 40; i++)
+            {
+                int featureId = i;
+                fetchTasks.Add(Task.Run(async () =>
+                {
+                    try
+                    {
+                        result = await client.GetAsync($"{_timeZoneServiceUrl}{featureId}?f=json");
```

Comment:
Since there are only a few of these, I suspect it would be much faster to download them all as one chunk. You can use the `/query` endpoint and set:`where`, `outFields` and `f` to get what you want. Something like this: `https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/World_Time_Zones/FeatureServer/0//query?where=1%3D1&outFields=*&f=json`

If you go straight to the endpoint, `https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/World_Time_Zones/FeatureServer/0/` you will get a UI that you can use to experiment with to get all the parameters set that you want. This might also be a way to speed up the geometry engine calls, you can make a request to the endpoint that says, give me back the timezone for this point. That query looks something like this: `https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/World_Time_Zones/FeatureServer/0//query?where=1%3D1&objectIds=&time=&geometry=-122.614408%2C45.5377835&geometryType=esriGeometryPoint&inSR=4326&spatialRel=esriSpatialRelIntersects&resultType=none&distance=0.0&units=esriSRUnit_Meter&relationParam=&returnGeodetic=false&outFields=*&returnGeometry=true&returnCentroid=false&featureEncoding=esriDefault&multipatchOption=xyFootprint&maxAllowableOffset=&geometryPrecision=&outSR=&defaultSR=&datumTransformation=&applyVCSProjection=false&returnIdsOnly=false&returnUniqueIdsOnly=false&returnCountOnly=false&returnExtentOnly=false&returnQueryGeometry=false&returnDistinctValues=false&cacheHint=false&orderByFields=&groupByFieldsForStatistics=&outStatistics=&having=&resultOffset=&resultRecordCount=&returnZ=false&returnM=false&returnExceededLimitFeatures=true&quantizationParameters=&sqlFormat=none&f=html&token=`

Open that link to see the HTML page with the parameters filled out. Basically that's an x/y location (in 4326, WGS86, latlongs) and it will return just a single polygon/timezone.

Reply:
I'll look into this later today.